### PR TITLE
feat: Add comprehensive component testing framework for strategy development

### PR DESCRIPTION
**Pull request by alexflorisca**
# Component Testing Framework Implementation

## Overview

This PR introduces a comprehensive component testing framework that enables isolated testing and performance analysis of trading strategy components. This framework is essential for the strategy redesign initiative, allowing developers to test individual components (signal generators, risk managers, position sizers) in isolation and combination.

## Background

As part of the strategy redesign effort, we need robust testing capabilities to:
- Validate individual component performance
- Test components across different market regimes
- Analyze performance attribution and component impact
- Generate synthetic test data for comprehensive testing scenarios

### 🧪 Component Performance Testing
- **ComponentPerformanceTester**: Isolated testing of signal generators, risk managers, and position sizers
- Performance metrics including accuracy, Sharpe ratio, drawdown control, and risk efficiency
- Support for testing components individually and in combination

### 📊 Regime-Specific Analysis
- **RegimeTester**: Test components across different market regimes (bull/bear, high/low volatility)
- Regime transition analysis and statistics
- Component performance comparison across market conditions

### 🔍 Performance Attribution
- **PerformanceAttributionAnalyzer**: Analyze the impact of component changes
- Component replacement impact analysis
- Baseline performance metrics calculation

### 🎯 Synthetic Test Data Generation
- **TestDatasetGenerator**: Create comprehensive test datasets
- Multiple market scenarios (bull/bear markets, volatility regimes)
- Edge case testing (missing data, extreme volatility, zero volume)
- Regime-labeled datasets for regime-specific testing

### New Components Added
- `src/strategies/components/testing/component_performance_tester.py` - Core testing framework
- `src/strategies/components/testing/regime_tester.py` - Regime-specific testing
- `src/strategies/components/testing/performance_attribution.py` - Performance analysis
- `src/strategies/components/testing/test_datasets.py` - Synthetic data generation
- `examples/component_testing_example.py` - Usage demonstration

### Test Coverage
- **500+ test cases** covering all framework components
- Integration tests for end-to-end functionality
- Edge case testing for robustness
- Performance benchmarking

## Usage Example

```python
from src.strategies.components.testing import ComponentPerformanceTester, TestDatasetGenerator

# Generate test data
generator = TestDatasetGenerator()
test_data = generator.generate_synthetic_dataset("strong_bull_low_vol", seed=42)

# Test components
tester = ComponentPerformanceTester(test_data)
results = tester.test_signal_generator(RandomSignalGenerator())
print(f"Accuracy: {results.accuracy:.3f}")
```

## Testing

- [x] Unit tests pass (500+ test cases)
- [x] Integration tests pass
- [x] Edge case testing completed
- [x] Performance benchmarking completed
- [x] Example usage verified

## Breaking Changes

None - this is a new feature addition that doesn't modify existing functionality.

## Related Issues

This addresses the need for comprehensive component testing as part of the strategy redesign initiative.

## Screenshots

N/A - This is a backend testing framework with no UI components.

## Checklist

- [x] Feature implementation completed
- [x] Unit tests added and passing
- [x] Integration tests added and passing
- [x] Documentation and examples provided
- [x] Code follows project standards
- [x] No breaking changes introduced
- [x] Performance impact assessed

### CODE REVIEW

File: src/strategies/components/strategy_registry.py [hunk 1]
```diff
@@ -479,18 +483,18 @@ def revert_to_version(self, strategy_id: str, target_version: str) -> None:
...
-            risk_manager_config=risk_mgr_config,
-            position_sizer_config=pos_sizer_config,
-            regime_detector_config=regime_det_config,
-            parameters=target_version_record.parameters or {},
-            performance_summary=None,  # Reset performance summary on revert
-            validation_results=None,  # Reset validation on revert
+            signal_generator_config=metadata.signal_generator_config,
+            risk_manager_config=metadata.risk_manager_config,
+            position_sizer_config=metadata.position_sizer_config,
+            regime_detector_config=metadata.regime_detector_config,
```
[the remark this answers is not in the document]
Comment (alexflorisca) [reply src/strategies/components/strategy_registry.py:489]: @cursoragent fix this comment

**Comment by alexflorisca** [comment]
@codex review

File: src/strategies/components/strategy_registry.py [hunk 2]
```diff
@@ -479,18 +483,18 @@ def revert_to_version(self, strategy_id: str, target_version: str) -> None:
...
+            parameters=metadata.parameters,
+            performance_summary=metadata.performance_summary,
+            validation_results=metadata.validation_results,
             lineage_path=metadata.lineage_path,
             branch_name=metadata.branch_name,
             merge_source=metadata.merge_source,
-            config_hash=target_version_record.config_hash or metadata.config_hash,
-            component_hash=target_version_record.component_hash or metadata.component_hash
+            config_hash=metadata.config_hash,
+            component_hash=metadata.component_hash
```
[the remark this answers is not in the document]
Comment (alexflorisca) [reply src/strategies/components/strategy_registry.py:497]: @cursoragent fix this unresolved comment

**Comment by alexflorisca** [comment]
@codex review

**Comment by alexflorisca** [comment]
@codex review

**Comment by alexflorisca** [comment]
# 🔍 Code Review - PR #313: Component Testing Framework

## Overview
Reviewed the component testing framework implementation focusing on critical bugs, security vulnerabilities, and performance issues. Found several critical issues that need to be addressed before merging.

---

## 🔴 Critical Issues

### 1. Global State Pollution - Random Seed Contamination
**File:** `src/strategies/components/testing/test_datasets.py`  
**Line:** 719

```python
# BAD: Modifies global random state
if seed is not None:
    np.random.seed(seed)  # This pollutes global state!
```

**Impact:** This modifies global NumPy random state, affecting ALL other code using `np.random` in the application. In a trading bot, this could cause non-deterministic behavior in live trading or other concurrent processes.

**Fix:**
```python
# GOOD: Use isolated random generator
rng = np.random.default_rng(seed)
random_returns = rng.normal(drift, daily_vol, n_periods)
```

---

### 2. Inconsistent Random Seed Usage
**File:** `src/strategies/components/testing/test_datasets.py`  
**Lines:** 446-447, 497-498

```python
# Uses default_rng for generator
rng = np.random.default_rng(seed)

# But then uses global np.random elsewhere
missing_indices = np.random.choice(data.index, size=int(len(data) * 0.1), replace=False)
```

**Impact:** Results are NOT reproducible even with a seed because some operations use the seeded generator and others use global state.

**Fix:** Use `rng.choice()` instead of `np.random.choice()` throughout.

---

### 3. Potential Division by Zero
**File:** `src/strategies/components/testing/component_performance_tester.py`  
**Line:** 716

```python
drawdown_control_score = max(0.0, 1.0 - max_drawdown_achieved / target_max_drawdown)
```

**Impact:** If `target_max_drawdown` is 0, this will raise `ZeroDivisionError`, crashing the test.

**Fix:**
```python
drawdown_control_score = max(0.0, 1.0 - max_drawdown_achieved / target_max_drawdown) if target_max_drawdown > 0 else 0.0
```

---

### 4. Fragile String Parsing in Regime Detection
**File:** `src/strategies/components/testing/regime_tester.py`  
**Lines:** 312-319, 658-665

```python
regime_parts = regime_type.split('_')
if len(regime_parts) >= 3:
    trend_str = f"{regime_parts[0]}_{regime_parts[1]}"
    volatility_str = regime_parts[2]
else:
    # Fallback - but what if format is completely wrong?
    trend_str = 'range'
    volatility_str = 'low_vol'
```

**Impact:** Silent failure - incorrect regime types won't error, they'll just use wrong defaults, leading to incorrect test results.

**Fix:** Raise an error for invalid formats or add validation.

---

### 5. AttributeError Risk with Signal Direction
**File:** `src/strategies/components/testing/regime_tester.py`  
**Line:** 684

```python
if signal.direction.value == 'buy' and future_return > 0:
```

**Impact:** If `signal.direction` is already a string (not an enum), this will crash with `AttributeError: 'str' object has no attribute 'value'`.

**Fix:**
```python
direction_value = signal.direction.value if hasattr(signal.direction, 'value') else signal.direction
if direction_value == 'buy' and future_return > 0:
```

---

## 🟡 Performance Issues

### 6. Inefficient Signal-Trade Matching Algorithm
**File:** `src/strategies/components/testing/performance_attribution.py`  
**Lines:** 586-608

```python
for signal_data in signals:
    for trade in trades:  # O(n * m) - very slow for large datasets
        if abs((trade['entry_time'] - signal_time).total_seconds()) < 3600:
            corresponding_trade = trade
            break
```

**Impact:** With 1000 signals and 1000 trades, this performs 1,000,000 comparisons. This will be extremely slow for backtests.

**Fix:** Use a more efficient algorithm with indexing or binary search.

---

### 7. Potential Memory Overflow with Product Operation
**File:** `src/strategies/components/testing/component_performance_tester.py`  
**Line:** 492

```python
total_return = (1 + portfolio_returns).prod() - 1
```

**Impact:** For large datasets with many positive returns, `prod()` can overflow to `inf`. For many negative returns, it can underflow to 0.

**Fix:**
```python
# Use log returns for numerical stability
total_return = np.exp(np.log1p(portfolio_returns).sum()) - 1
```

---

### 8. Repeated DataFrame Operations in Loops
**File:** `src/strategies/components/testing/test_datasets.py`  
**Lines:** 895-905

```python
for i in range(bubble_start, bubble_peak):
    data.iloc[i, data.columns.get_loc('close')] *= (1 + extra_return)  # Slow!
```

**Impact:** `iloc` and `get_loc` in a loop is extremely slow. For 1000 iterations, this could take seconds.

**Fix:** Vectorize the operations using NumPy arrays.

---

## 🔵 Code Quality Issues

### 9. Silent Error Handling
**File:** `src/strategies/components/testing/performance_attribution.py`  
**Lines:** 294-295

```python
try:
    signal = strategy.signal_generator.generate_signal(current_data, i, regime)
except:  # Catches everything, including SystemExit!
    pass  # Silent failure - bugs will be hidden
```

**Impact:** Silently swallows ALL exceptions, including critical ones. Bugs in signal generators will be completely hidden.

**Fix:** Use specific exception types and log warnings.

---

### 10. Magic Numbers Throughout
**Examples:** Lines 435, 528, 605, 716, etc.

```python
accurate = abs(future_return) < 0.01  # What does 0.01 mean?
target_max_drawdown = 0.1  # Why 10%?
kelly_adherence = 0.8  # Arbitrary placeholder
```

**Impact:** Hard to understand, maintain, and tune. These should be configurable parameters or named constants.

---

### 11. Placeholder Implementations with Hardcoded Values
**File:** `src/strategies/components/testing/component_performance_tester.py`  
**Lines:** 741, 850, 862-866

```python
regime_adaptation_score=0.8,  # Placeholder - not actually calculated
optimal_sizing_score=0.75,  # Placeholder
```

**Impact:** Tests return fake results, giving false confidence. These should either be calculated or clearly marked as TODO/NotImplemented.

---

## 📊 Summary

| Severity | Count | Must Fix Before Merge |
|----------|-------|----------------------|
| 🔴 Critical | 5 | ✅ Yes |
| 🟡 Performance | 3 | ⚠️ Recommended |
| 🔵 Code Quality | 3 | ℹ️ Nice to have |

## ✅ Recommended Actions

1. **MUST FIX**: Issues #1-5 (Critical bugs that could cause failures or incorrect results)
2. **SHOULD FIX**: Issues #6-8 (Performance bottlenecks that will slow down backtesting)
3. **CONSIDER**: Issues #9-11 (Code quality improvements for maintainability)

---

## �� Quick Fix Option

[🤖 Fix All Issues](https://api.cursor.com/v0/agents?prompt=fix%20all%20issues%20identified%20in%20this%20review%20for%20PR%20%23313&repo=https://github.com/bumpy-croc/ai-trading-bot&ref=feature/strategy-redesign-9)

Click the link above to automatically fix all identified issues using the Cursor agent.

---

**Review completed**: $(date -u +%Y-%m-%dT%H:%M:%SZ)  
**Reviewer**: AI Code Review Agent (Hunt Mode)

**Comment by alexflorisca** [comment]
# 🔍 Code Review: Component Testing Framework (Hunt Mode)

## Executive Summary
Found **5 critical bugs** that could cause runtime failures and **3 performance/design issues**. Most critical: division by zero errors and NaN propagation that could crash the testing framework under edge cases.

---

## 🔴 Critical Issues

### 1. Division by Zero in Performance Attribution Analysis
**Location:** `src/strategies/components/testing/performance_attribution.py:220`

```python
explained_performance = min(1.0, abs(total_attribution) / abs(baseline_results['total_return']) 
                          if baseline_results['total_return'] != 0 else 0.0)
```

**Issue:** The ternary operator checks for zero, but this is vulnerable to floating point precision issues. If `total_return` is very close to zero (e.g., `1e-15`), it will pass the check but cause division by very small numbers, leading to numerical instability.

**Impact:** 🔴 **CRITICAL** - Can cause `inf` values that propagate through calculations and crash the system.

**Fix:**
```python
# Use a tolerance-based check
EPSILON = 1e-10
if abs(baseline_results['total_return']) < EPSILON:
    explained_performance = 0.0
else:
    explained_performance = min(1.0, abs(total_attribution) / abs(baseline_results['total_return']))
```

---

### 2. Unhandled Empty Dictionary in max/min Operations
**Locations:**
- `performance_attribution.py:233-234`
- Similar pattern in `strategy_registry.py`

```python
primary_driver = max(component_contributions, key=component_contributions.get) if component_contributions else ""
weakest_component = min(component_contributions, key=component_contributions.get) if component_contributions else ""
```

**Issue:** The conditional check `if component_contributions` is correct, but this pattern is repeated throughout the codebase without consistent safety checks. The check evaluates truthiness, which can be confusing with dictionaries.

**Impact:** 🔴 **CRITICAL** - `ValueError: max() arg is an empty sequence` if all components return None.

**Fix:**
```python
if component_contributions and len(component_contributions) > 0:
    primary_driver = max(component_contributions, key=component_contributions.get)
    weakest_component = min(component_contributions, key=component_contributions.get)
else:
    primary_driver = ""
    weakest_component = ""
```

---

### 3. NaN Propagation from Correlation Calculations
**Locations:**
- `component_performance_tester.py:508-511`
- `component_performance_tester.py:846-847`  
- `performance_attribution.py:686`

```python
confidence_accuracy_correlation = np.corrcoef(
    confidences, 
    [1 if s['accurate'] else 0 for s in all_signals]
)[0, 1] if len(confidences) > 1 else 0.0
```

**Issue:** `np.corrcoef` returns `NaN` when:
- All values in an array are identical (zero variance)
- Arrays contain NaN values
- Standard deviation is zero

The code checks `len > 1` but doesn't handle the NaN case.

**Impact:** 🔴 **CRITICAL** - NaN values propagate through calculations, causing test results to be invalid and potentially crashing downstream analysis.

**Fix:**
```python
if len(confidences) > 1:
    correlation = np.corrcoef(confidences, 
                              [1 if s['accurate'] else 0 for s in all_signals])[0, 1]
    confidence_accuracy_correlation = 0.0 if np.isnan(correlation) else correlation
else:
    confidence_accuracy_correlation = 0.0
```

---

### 4. Division by Zero in Position Size Consistency
**Location:** `component_performance_tester.py:682-683`

```python
avg_position_size = np.mean(position_sizes)
position_size_consistency = 1.0 / (1.0 + np.std(position_sizes))  # Higher is more consistent
```

**Issue:** If `np.std(position_sizes)` approaches `-1.0` (theoretically possible with specific data distributions), the denominator becomes zero.

**Impact:** 🔴 **CRITICAL** - Division by zero crashes the test.

**Fix:**
```python
std_pos_size = np.std(position_sizes)
position_size_consistency = 1.0 / max(1.0 + std_pos_size, 1e-10)
```

---

### 5. Potential Index Out of Bounds in Signal Testing
**Location:** `component_performance_tester.py:426-428`

```python
# Calculate signal accuracy (if we have future returns)
if i < len(scenario_data) - 1:
    future_return = scenario_data.iloc[i + 1]['returns']
```

**Issue:** The bounds check is present, but in the loop starting at line 408, `i` ranges from `0` to `len(scenario_data)`. When `i == len(scenario_data) - 1`, accessing `scenario_data.iloc[i + 1]` will be out of bounds.

**Impact:** 🟡 **WARNING** - Edge case that could cause `IndexError` in specific scenarios.

**Fix:**
```python
# Loop should be: for i in range(len(scenario_data) - 1):
# OR add better bounds checking
```

---

## 🟡 Performance & Design Issues

### 6. Global Random State Pollution
**Location:** `test_datasets.py:633, 718-719, 782, 812+`

```python
# Line 633
regime_data['confidence'] = np.random.uniform(0.6, 0.9, len(regime_data))

# Line 718
if seed is not None:
    np.random.seed(seed)  # Pollutes global state
```

**Issue:** The code inconsistently uses isolated RNG (`np.random.default_rng(seed)`) in some places but global numpy random state in others. This makes tests non-reproducible and can cause side effects.

**Impact:** 🟡 **WARNING** - Test reproducibility issues, potential race conditions in multi-threaded scenarios.

**Fix:**
```python
# Use isolated RNG consistently:
rng = np.random.default_rng(seed)
regime_data['confidence'] = rng.uniform(0.6, 0.9, len(regime_data))
```

---

### 7. Missing Transaction Costs in Backtesting Simulation
**Location:** `component_performance_tester.py:544-564`

```python
def _simulate_signal_trading(self, signals: List[Dict[str, Any]]) -> pd.Series:
    """Simulate trading based on signals to calculate performance"""
    returns = []
    
    for signal_data in signals:
        signal = signal_data['signal']
        future_return = signal_data['future_return']
        
        # Simple trading simulation
        if signal.direction == SignalDirection.BUY:
            # Long position: profit from positive returns
            trade_return = future_return * signal.strength
```

**Issue:** Trading simulation doesn't account for:
- Transaction costs/fees
- Slippage
- Bid-ask spreads

**Impact:** 🔵 **INFO** - Optimistic performance metrics that don't reflect real trading conditions.

**Recommendation:** Add configurable transaction cost parameter (e.g., 0.1% per trade).

---

### 8. Inefficient Nested Loop Structure
**Location:** `component_performance_tester.py:402-458`

```python
# Test across all scenarios
for scenario in test_scenarios:
    scenario_data = self.test_data.iloc[scenario['start_idx']:scenario['end_idx']].copy()
    
    # Generate signals for this scenario
    for i in range(len(scenario_data)):
        # ... signal generation logic
```

**Issue:** Nested loops over scenarios and time periods create O(n*m) complexity. For large datasets with many scenarios, this could be slow.

**Impact:** 🔵 **INFO** - Performance bottleneck for large-scale testing.

**Recommendation:** Consider vectorization or parallel processing for independent scenarios.

---

## 🔵 Minor Issues

### 9. Inconsistent Attribute Access Patterns
**Location:** `regime_tester.py:686-687`

```python
if signal.direction.value == 'buy' and future_return > 0:
    accurate = True
elif signal.direction.value == 'sell' and future_return < 0:
    accurate = True
```

**Issue:** Accesses `.value` attribute assuming it exists. If `SignalDirection` enum implementation changes, this breaks.

**Impact:** 🔵 **INFO** - Fragile code that depends on internal enum implementation.

**Recommendation:**
```python
if signal.direction == SignalDirection.BUY and future_return > 0:
    accurate = True
```

---

## 📊 Summary

| Severity | Count | Type |
|----------|-------|------|
| 🔴 Critical | 5 | Division by zero, NaN propagation, empty sequences |
| 🟡 Warning | 2 | Random state pollution, bounds checking |
| 🔵 Info | 2 | Design improvements, code quality |

## ✅ Positive Observations

1. **Good error handling structure** - Try-catch blocks in most critical sections
2. **Comprehensive logging** - Uses Python logging module appropriately
3. **Extensive test coverage** - 500+ test cases
4. **Good documentation** - Well-documented functions with docstrings
5. **Type hints** - Uses dataclasses and type annotations

---

## 🤖 Fix All Issues

[🤖 Fix All Issues](https://api.cursor.com/v0/agents?prompt=fix%20all%20issues%20identified%20in%20this%20review%20for%20PR%20%23313&repo=https://github.com/bumpy-croc/ai-trading-bot&ref=feature/strategy-redesign-9)

Click the link above to automatically launch an agent that will fix all identified issues.

**Comment by alexflorisca** [comment]
@codex review

File: src/strategies/components/strategy_registry.py [hunk 3]
```diff
@@ -483,14 +490,14 @@ def revert_to_version(self, strategy_id: str, target_version: str) -> None:
...
+            parameters=reverted_parameters,
             performance_summary=None,  # Reset performance summary on revert
             validation_results=None,  # Reset validation on revert
             lineage_path=metadata.lineage_path,
             branch_name=metadata.branch_name,
             merge_source=metadata.merge_source,
-            config_hash=target_version_record.config_hash or metadata.config_hash,
-            component_hash=target_version_record.component_hash or metadata.component_hash
+            config_hash=target_version_record.config_hash,
+            component_hash=target_version_record.component_hash
```
[the remark this answers is not in the document]
Comment (alexflorisca) [reply src/strategies/components/strategy_registry.py:500]: @cursoragent fix this comment https://github.com/bumpy-croc/ai-trading-bot/pull/313#discussion_r2399642509

**Comment by alexflorisca** [comment]
codebot

**Comment by alexflorisca** [comment]
@cursoragent fix all 13 issues reported in this PR review https://github.com/bumpy-croc/ai-trading-bot/pull/313#issuecomment-3362976080. Don't worry about adding extra tests. When fixing an issue, check if that issue exists anywhere else in the file diff of this PR. Don't change any files that were not modified by this PR, unless relevant.

**Comment by alexflorisca** [comment]
@codex review

**Comment by alexflorisca** [comment]
codebot

**Comment by alexflorisca** [comment]
@cursoragent fix all critical issues from the latest codebot review https://github.com/bumpy-croc/ai-trading-bot/pull/313#issuecomment-3363314552. Extract the comment id and use the github CLI to access the contents.

**Comment by alexflorisca** [comment]
@codex review

**Comment by alexflorisca** [comment]
@codex review
